### ISSUE TITLE
Get-AzureRmRoleAssignment failing with "Object reference not set to an instance of an object"

### ISSUE THREAD

**Issue by paulomarquesc** [Azure/azure-powershell#4828]
### Cmdlet(s)
Get-AzureRmRoleAssignment

### PowerShell Version
5.1.15063.608

### Module Version
4.4.1

### OS Version
10.0.15063.608

### Description
This cmdlet is failing when you use -SignInName or -ServicePrincipalName parameters with the following error message:

Get-AzureRmRoleAssignment : Object reference not set to an instance of an object.
At line:1 char:1
+ Get-AzureRmRoleAssignment -Scope $scope -SignInName "paulo@pmcglobal. ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Get-AzureRmRoleAssignment], NullReferenceException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand

### Debug Output
DEBUG: 8:34:53 AM - GetAzureRoleAssignmentCommand begin processing with ParameterSet 'ScopeWithSignInNameParameterSet'.
DEBUG: 8:34:53 AM - using account id 'admin01@pmcglobal.onmicrosoft.com'...
DEBUG: [Common.Authentication]: Authenticating using Account: 'admin01@pmcglobal.onmicrosoft.com', environment: 'AzureCloud', tenant: 'de3c01ef-53d2-477f-81e8-971ddc151d92'
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: f1be45b3-3b62-4514-8591-55053282da9b - AcquireTokenHandlerBase: === Token Acquisition started:
 Authority: https://login.microsoftonline.com/de3c01ef-53d2-477f-81e8-971ddc151d92/
 Resource: https://graph.windows.net/
 ClientId: 1950a258-227b-4e31-a9cf-717495945fc2
 CacheType: Microsoft.Azure.Commands.Common.Authentication.AuthenticationStoreTokenCache (2 items)
 Authentication Target: User

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/21/2017 15:34:53: f1be45b3-3b62-4514-8591-55053282da9b - TokenCache: Looking up cache for a token...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: f1be45b3-3b62-4514-8591-55053282da9b - TokenCache: An item matching the requested resource was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/21/2017 15:34:53: f1be45b3-3b62-4514-8591-55053282da9b - TokenCache: 49.3084799416667 minutes left until token in cache expires
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: f1be45b3-3b62-4514-8591-55053282da9b - TokenCache: A matching item (access token or refresh token or both) was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: f1be45b3-3b62-4514-8591-55053282da9b - AcquireTokenHandlerBase: === Token Acquisition finished successfully. An access token was retuned:
 Access Token Hash: MtlcSWlDlzsh64imKUJ5bXYTbrHmvdlTxQTw6MXIzZY=
 Refresh Token Hash: 7OLsesXamvyCPTEIKomTJWDq+sV4JbSHpgzMQQDQL1M=
 Expiration Time: 10/21/2017 16:24:12 +00:00
 User Hash: KU4QzWkkDwgn8U3MAUDec5GltjdmpyWzVy+c5RL80MY=

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53:  - TokenCache: Serializing token cache with 2 items.
DEBUG: [Common.Authentication]: Authenticating using Account: 'admin01@pmcglobal.onmicrosoft.com', environment: 'AzureCloud', tenant: 'de3c01ef-53d2-477f-81e8-971ddc151d92'
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 15e3c412-47b4-46c2-a098-502b0a9f0c30 - AcquireTokenHandlerBase: === Token Acquisition started:
 Authority: https://login.microsoftonline.com/de3c01ef-53d2-477f-81e8-971ddc151d92/
 Resource: https://management.core.windows.net/
 ClientId: 1950a258-227b-4e31-a9cf-717495945fc2
 CacheType: Microsoft.Azure.Commands.Common.Authentication.AuthenticationStoreTokenCache (2 items)
 Authentication Target: User

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/21/2017 15:34:53: 15e3c412-47b4-46c2-a098-502b0a9f0c30 - TokenCache: Looking up cache for a token...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 15e3c412-47b4-46c2-a098-502b0a9f0c30 - TokenCache: An item matching the requested resource was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/21/2017 15:34:53: 15e3c412-47b4-46c2-a098-502b0a9f0c30 - TokenCache: 46.3204760533333 minutes left until token in cache expires
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 15e3c412-47b4-46c2-a098-502b0a9f0c30 - TokenCache: A matching item (access token or refresh token or both) was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 15e3c412-47b4-46c2-a098-502b0a9f0c30 - AcquireTokenHandlerBase: === Token Acquisition finished successfully. An access token was retuned:
 Access Token Hash: /7gSBHIxp1R18/g+F6tiKdoPcLlFwY0rz6eG71cQsg0=
 Refresh Token Hash: 7OLsesXamvyCPTEIKomTJWDq+sV4JbSHpgzMQQDQL1M=
 Expiration Time: 10/21/2017 16:21:13 +00:00
 User Hash: KU4QzWkkDwgn8U3MAUDec5GltjdmpyWzVy+c5RL80MY=

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53:  - TokenCache: Serializing token cache with 2 items.
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 84f615fb-a1a8-4d9f-8638-970171f7e364 - AcquireTokenHandlerBase: === Token Acquisition started:
 Authority: https://login.microsoftonline.com/de3c01ef-53d2-477f-81e8-971ddc151d92/
 Resource: https://graph.windows.net/
 ClientId: 1950a258-227b-4e31-a9cf-717495945fc2
 CacheType: Microsoft.Azure.Commands.Common.Authentication.AuthenticationStoreTokenCache (2 items)
 Authentication Target: User

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/21/2017 15:34:53: 84f615fb-a1a8-4d9f-8638-970171f7e364 - TokenCache: Looking up cache for a token...
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 84f615fb-a1a8-4d9f-8638-970171f7e364 - TokenCache: An item matching the requested resource was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Verbose: 1 :
DEBUG: 10/21/2017 15:34:53: 84f615fb-a1a8-4d9f-8638-970171f7e364 - TokenCache: 49.3084047416667 minutes left until token in cache expires
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 84f615fb-a1a8-4d9f-8638-970171f7e364 - TokenCache: A matching item (access token or refresh token or both) was found in the cache
DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53: 84f615fb-a1a8-4d9f-8638-970171f7e364 - AcquireTokenHandlerBase: === Token Acquisition finished successfully. An access token was retuned:
 Access Token Hash: MtlcSWlDlzsh64imKUJ5bXYTbrHmvdlTxQTw6MXIzZY=
 Refresh Token Hash: 7OLsesXamvyCPTEIKomTJWDq+sV4JbSHpgzMQQDQL1M=
 Expiration Time: 10/21/2017 16:24:12 +00:00
 User Hash: KU4QzWkkDwgn8U3MAUDec5GltjdmpyWzVy+c5RL80MY=

DEBUG: Microsoft.IdentityModel.Clients.ActiveDirectory Information: 2 :
DEBUG: 10/21/2017 15:34:53:  - TokenCache: Serializing token cache with 2 items.
DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
GET

Absolute Uri:
https://graph.windows.net/de3c01ef-53d2-477f-81e8-971ddc151d92/users?$filter=userPrincipalName eq 'paulo@pmcglobal.me'&api-version=1.6

Headers:
x-ms-client-request-id        : 59fca079-058e-4f62-bd0d-39d73f6bbafc
accept-language               : en-US

Body:

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
OK

Headers:
Pragma                        : no-cache
ocp-aad-diagnostics-server-name: DtvcVPtqaaHXo1s0keQF5e5ePpB7wWGVxIlKdtjGXok=
request-id                    : e1267bff-7883-4849-a2b8-024364ed164a
client-request-id             : fab4840b-fcfd-42ee-bb87-7a53029278ff
x-ms-dirapi-data-contract-version: 1.6
ocp-aad-session-key           : Bw5OWw5eGPWF-UBQe2Dcl7Wthm5NziEV94fGaqTnT8dHonA9pximF-dYicDho4a5vNGpKAqXGXiRw0NYL5xAK_ZoD1PKnyC-R5gk2GV5BbZXMFLlTFkS8-XdRHjNybZF.P0gPSA50BoTa6-53K4uub0DaRAC3uPk4LwTBwagLlB0
X-Content-Type-Options        : nosniff
DataServiceVersion            : 3.0;
Strict-Transport-Security     : max-age=31536000; includeSubDomains
Access-Control-Allow-Origin   : *
Duration                      : 1003746
Cache-Control                 : no-cache
Server                        : Microsoft-IIS/8.5
X-AspNet-Version              : 4.0.30319
X-Powered-By                  : ASP.NET,ASP.NET
Date                          : Sat, 21 Oct 2017 15:34:52 GMT

Body:
{
  "odata.metadata": "https://graph.windows.net/de3c01ef-53d2-477f-81e8-971ddc151d92/$metadata#directoryObjects/Microsoft.DirectoryServices.User",
  "value": [
    {
      "odata.type": "Microsoft.DirectoryServices.User",
      "objectType": "User",
      "objectId": "4a277c40-01ab-4ce1-8409-0c8bc4b7a92f",
      "deletionTimestamp": null,
      "accountEnabled": true,
      "assignedLicenses": [
        {
          "disabledPlans": [],
          "skuId": "a403ebcc-fae0-4ca2-8c8c-7a907fd6c235"
        },
        {
          "disabledPlans": [],
          "skuId": "8c4ce438-32a7-4ac5-91a6-e22ae08d9c8b"
        },
        {
          "disabledPlans": [],
          "skuId": "189a915c-fe4f-4ffa-bde4-85b9628d07a0"
        }
      ],
      "assignedPlans": [
        {
          "assignedTimestamp": "2017-09-18T00:15:13Z",
          "capabilityStatus": "Enabled",
          "service": "RMSOnline",
          "servicePlanId": "7a39d7dd-e456-4e09-842a-0204ee08187b"
        },
        {
          "assignedTimestamp": "2017-09-18T00:15:13Z",
          "capabilityStatus": "Enabled",
          "service": "AzureAnalysis",
          "servicePlanId": "2049e525-b859-401b-b2a0-e0a31c4b1fe4"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "SharePoint",
          "servicePlanId": "a361d6e2-509e-4e25-a8ad-950060064ef4"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "SharePoint",
          "servicePlanId": "527f7cdd-0e86-4c47-b879-f5fd357a3ac6"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "MicrosoftStream",
          "servicePlanId": "6c6042f5-6f01-4d67-b8c1-eb99d36eed3e"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "OfficeForms",
          "servicePlanId": "e212cbc7-0961-4c40-9825-01117710dcb1"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "ProcessSimple",
          "servicePlanId": "76846ad7-7776-4c40-a281-a386362dd1b9"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "PowerAppsService",
          "servicePlanId": "c68f8d98-5534-41c8-bf36-22fa496fa792"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "TeamspaceAPI",
          "servicePlanId": "57ff2da0-773e-42df-b2af-ffb7a2317929"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "ProjectWorkManagement",
          "servicePlanId": "b737dad2-2f6c-4c65-90e3-ca563267e8b9"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "Sway",
          "servicePlanId": "a23b959c-7ce8-4e57-9140-b90eb88a9e97"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "MicrosoftOffice",
          "servicePlanId": "43de0ff5-c92c-492b-9116-175376d08c38"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "MicrosoftCommunicationsOnline",
          "servicePlanId": "0feaeb32-d00e-4d66-bd5a-43b5b83db82c"
        },
        {
          "assignedTimestamp": "2017-08-23T16:26:09Z",
          "capabilityStatus": "Enabled",
          "service": "exchange",
          "servicePlanId": "efb87545-963c-4e0d-99df-69c6916d9eb0"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:41Z",
          "capabilityStatus": "Deleted",
          "service": "AADPremiumService",
          "servicePlanId": "eec0eb4f-6444-4f95-aba0-50c24d67f998"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "SCO",
          "servicePlanId": "c1ec4a95-1f05-45b3-a911-aa3fa01094f5"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "MultiFactorService",
          "servicePlanId": "8a256a2b-b617-496d-b51b-e76466e88db0"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "RMSOnline",
          "servicePlanId": "6c57d4b6-3b23-47a5-9bc9-69f17b4947b3"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "RMSOnline",
          "servicePlanId": "bea4c11e-220a-4e6d-8eb8-8ea15d019f90"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "AADPremiumService",
          "servicePlanId": "41781fb2-bc02-4b7c-bd55-b576c07bb09d"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "Deskless",
          "servicePlanId": "8c7d2df8-86f0-4902-b2ed-a0458298f3b3"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "MicrosoftStream",
          "servicePlanId": "9e700747-8b1d-45e5-ab8d-ef187ceec156"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "OfficeForms",
          "servicePlanId": "2789c901-c14e-48ab-a76a-be334d9d793a"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:41Z",
          "capabilityStatus": "Deleted",
          "service": "SharePoint",
          "servicePlanId": "5dbe027f-2339-4123-9542-606e4d348a72"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:41Z",
          "capabilityStatus": "Deleted",
          "service": "SharePoint",
          "servicePlanId": "e95bec33-7c88-4a70-8e19-b10bd9d0c014"
        },
        {
          "assignedTimestamp": "2017-08-23T16:24:40Z",
          "capabilityStatus": "Deleted",
          "service": "YammerEnterprise",
          "servicePlanId": "7547a3fe-08ee-4ccb-b430-5077c5041653"
        }
      ],
      "city": null,
      "companyName": null,
      "country": null,
      "creationType": null,
      "department": null,
      "dirSyncEnabled": null,
      "displayName": "Paulo Marques da Costa",
      "employeeId": null,
      "facsimileTelephoneNumber": null,
      "givenName": "Paulo",
      "immutableId": null,
      "isCompromised": null,
      "jobTitle": null,
      "lastDirSyncTime": null,
      "mail": "paulo@pmcglobal.me",
      "mailNickname": "paulo",
      "mobile": null,
      "onPremisesDistinguishedName": null,
      "onPremisesSecurityIdentifier": null,
      "otherMails": [],
      "passwordPolicies": "None",
      "passwordProfile": null,
      "physicalDeliveryOfficeName": null,
      "postalCode": null,
      "preferredLanguage": null,
      "provisionedPlans": [
        {
          "capabilityStatus": "Deleted",
          "provisioningStatus": "Success",
          "service": "SharePoint"
        },
        {
          "capabilityStatus": "Enabled",
          "provisioningStatus": "Success",
          "service": "SharePoint"
        },
        {
          "capabilityStatus": "Enabled",
          "provisioningStatus": "Success",
          "service": "SharePoint"
        },
        {
          "capabilityStatus": "Deleted",
          "provisioningStatus": "Success",
          "service": "SharePoint"
        },
        {
          "capabilityStatus": "Enabled",
          "provisioningStatus": "Success",
          "service": "MicrosoftCommunicationsOnline"
        },
        {
          "capabilityStatus": "Enabled",
          "provisioningStatus": "Success",
          "service": "exchange"
        }
      ],
      "provisioningErrors": [],
      "proxyAddresses": [
        "SMTP:paulo@pmcglobal.me"
      ],
      "refreshTokensValidFromDateTime": "2017-07-03T18:14:47Z",
      "showInAddressList": null,
      "signInNames": [],
      "sipProxyAddress": "paulo@pmcglobal.me",
      "state": null,
      "streetAddress": null,
      "surname": "Marques da Costa",
      "telephoneNumber": null,
      "usageLocation": "US",
      "userIdentities": [],
      "userPrincipalName": "paulo@pmcglobal.me",
      "userType": "Member"
    }
  ]
}

Get-AzureRmRoleAssignment : Object reference not set to an instance of an object.
At line:1 char:1
+ Get-AzureRmRoleAssignment -Scope $scope -SignInName "paulo@pmcglobal. ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Get-AzureRmRoleAssignment], NullReferenceException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand

DEBUG: AzureQoSEvent: CommandName - Get-AzureRmRoleAssignment; IsSuccess - False; Duration - 00:00:00.3873269; Exception - System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Azure.Commands.Resources.Models.Authorization.AuthorizationClient.FilterRoleAssignments(FilterRoleAssignmentsOptions options, String currentSubscription)
   at Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand.ExecuteCmdlet()
   at Microsoft.WindowsAzure.Commands.Utilities.Common.AzurePSCmdlet.ProcessRecord();
DEBUG: Finish sending metric.
DEBUG: 8:34:54 AM - GetAzureRoleAssignmentCommand end processing.
DEBUG: 8:34:54 AM - GetAzureRoleAssignmentCommand end processing.

### Script/Steps for Reproduction
1) Authenticate with "Add-AzureRmAccount"
2) Copy the subscription Id that will be used on step 3
3) Set a variable called $scope with the following command: $scope = "/subscriptions/<yoursubscription from step 2>"
4) Get the list of all your role assignments: $roleAssignments = Get-AzureRmRoleAssignment -Scope $scope
5) Look for one of the items that has a UPN associated (in my case item 2 of my list has this content)
6) Run: Get-AzureRmRoleAssignment -Scope $scope -SignInName $roleAssignments[1].SignInName

Then you get the error:
Get-AzureRmRoleAssignment : Object reference not set to an instance of an object.
At line:1 char:1
+ Get-AzureRmRoleAssignment -Scope $scope -SignInName $roleAssignments[ ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Get-AzureRmRoleAssignment], NullReferenceException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Resources.GetAzureRoleAssignmentCommand

This happens when using -ServicePrincipalName parameter as well.

**Comment by markcowl** [comment]
@darshanhs90 Can you take a look?

**Comment by paulomarquesc** [comment]
I just found out where the bug is, I debugged the solution and found out that the line 222 on file **azure-powershell\src\ResourceManager\Resources\Commands.Resources\Models.Authorization\AuthorizationClient.cs** is executing an IsNullOrEmpty in an object that is executing the ToString() method and when the object is null, the ToString() triggers a NullReferenceExcepytion.

The line is:
principalId = string.IsNullOrEmpty(options.ADObjectFilter.Id.ToString()) ? adObject.Id.ToString() : options.ADObjectFilter.Id;

and should be replaced by:
principalId = string.IsNullOrEmpty(options.ADObjectFilter.Id) ? adObject.Id.ToString() : options.ADObjectFilter.Id

This will give the result we need, which is the object Id, the string.IsNullOrEmpty(options.ADObjectFilter.Id) will be null and then the content of adObject.Id.ToString() will be returned.

I'll place a merge request with the fix.
